File: Prefix/Java/Count Vowel Strings in Ranges.java
// Problem link: https://leetcode.com/problems/count-vowel-strings-in-ranges/description
// Time: O(m + n)
// Space: O(m)
class Solution {
    public int[] vowelStrings(String[] words, int[][] queries) {
        int left, right, prefix[] = new int[words.length + 1], result[] = new int[queries.length];
        String vowels = "aeiou", word;

        for(int index = 0; index < words.length; index++) {
            word = words[index];
            prefix[index + 1] = prefix[index];
            if(vowels.indexOf(word.charAt(0)) != -1 && vowels.indexOf(word.charAt(word.length() - 1)) != -1) {
                prefix[index + 1]++;
            }
        }

        for(int index = 0; index < queries.length; index++) {
            left = queries[index][0];
            right = queries[index][1];
            result[index] = prefix[right + 1] - prefix[left];
        }

        return result;
    }
}

File: Prefix/Python/Count Vowel Strings in Ranges.py
# Problem link: https://leetcode.com/problems/count-vowel-strings-in-ranges/description
# Time: O(m + n)
# Space: O(m)
class Solution:
    def vowelStrings(self, words: List[str], queries: List[List[int]]) -> List[int]:
        prefix, vowels, result = [0], ['a', 'e', 'i', 'o', 'u'], []

        for word in words:
            prefix.append(prefix[-1])
            if word[0] in vowels and word[-1] in vowels:
                prefix[-1] += 1
                
        for start, end in queries:
            result.append(prefix[end + 1] - prefix[start])
        
        return result
